### PR TITLE
Disallow JavaScriptSerializer deserialization when initialized with SimpleTypeResolver

### PR DESCRIPTION
**Pull request by dotpaul**

### CODE REVIEW

**Comment by dotpaul** [comment]
FYI @LingxiaChen @timhermann @matt-desai @DSIE63

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                                        }
+                                        else
+                                        {
+                                            kind = PropertySetAbstractValueKind.Unflagged;
+                                        }
+
+                                        break;
+
+                                    default:
+                                        Debug.Fail($"Unhandled PointsToAbstractValueKind {pointsTo.Kind}");
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs]: So you don't expect other kinds, such as `PointsToAbstractValueKind.Unknown`? Will that not be common in presence of conservative analysis where PointsToAnalysis is unable to figure out the points to value and conservatively assumes unknown.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolverTests.cs [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+    {
+        public static JavaScriptSerializer Serializer;
+
+        public T D<T>(string str)
+        {
+            return (T) Program.Serializer.Deserialize(str, typeof(T));
+        }
+    }
+}",
+                GetCSharpResultAt(13, 24, MaybeRule, "object JavaScriptSerializer.Deserialize(string input, Type targetType)"));
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolverTests.cs]: So we flag a may be diagnostic even if we have no knowledge about the serializer?

Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolverTests.cs]: Yep.  That's configurable, by disabling the rule.  😁

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs (hunk 1 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs]: Hrm, I need more test cases for this.

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                "CA2321",
+                nameof(MicrosoftNetCoreSecurityResources.JavaScriptSerializerWithSimpleTypeResolverTitle),
+                nameof(MicrosoftNetCoreSecurityResources.JavaScriptSerializerWithSimpleTypeResolverMessage),
+                isEnabledByDefault: false,
+                helpLinkUri: null);
+        internal static readonly DiagnosticDescriptor MaybeWithSimpleTypeResolver =
+            SecurityHelpers.CreateDiagnosticDescriptor(
+                "CA2322",
+                nameof(MicrosoftNetCoreSecurityResources.JavaScriptSerializerWithSimpleTypeResolverTitle),
+                nameof(MicrosoftNetCoreSecurityResources.JavaScriptSerializerWithSimpleTypeResolverMessage),
```
Comment (LingxiaChen) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs]: Replace it with JavaScriptSerializerMaybeWithSimpleTypeResolverMessage?

Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureDeserializerJavascriptSerializerWithSimpleTypeResolver.cs]: Oh gosh, thanks!

File: src/Test.Utilities/DiagnosticExtensions.cs [hunk 4]
```diff
@@ -25,10 +27,12 @@ public static class DiagnosticExtensions
                 CompilationUtils.ValidateNoCompileErrors(diagnostics);
             }
 
-            var analyzerDiagnostics = compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync().Result;
```
Comment (dotpaul) [inline src/Test.Utilities/DiagnosticExtensions.cs]: Calling both GetAnalyzerDiagnosticsAsync() and GetAllDiagnosticsAsync() seemed to cause the weirdness I saw, where my analyzer's compilation end action and operation callbacks would be invoked twice with different Compilation objects, but the compilation start action would only be invoked once.

Comment (dotpaul) [reply src/Test.Utilities/DiagnosticExtensions.cs]: Opened https://github.com/dotnet/roslyn/issues/35149

**Comment by dotpaul** [comment]
Thank you @mavasani and @LingxiaChen !

**Pull request merged**